numpyro/__init__.py

```diff
@@ -1,6 +1,21 @@
 import numpyro.patch  # noqa: F401
-from numpyro.primitives import module, param, plate, sample  # noqa: F401
-from numpyro.version import __version__  # noqa: F401
-import numpyro.util as util
+from numpyro import compat, diagnostics, distributions, infer, infer_util, util
+from numpyro.primitives import module, param, plate, sample
+from numpyro.version import __version__
 
 util.set_platform('cpu')
+
+
+__all__ = [
+    '__version__',
+    'compat',
+    'diagnostics',
+    'distributions',
+    'infer',
+    'infer_util',
```

Comment:
should we move `infer_util` (and rename it to `util`) and `hmc_util` to `infer`?

Reply:
It is nice to be able to distinguish between utilities for setting random number generator seed and control flow primitives, from inference utilities for prediction and log density computation. I don't think most users of the library should care about anything in `hmc_util`, and if they do, it is probably worth exposing it directly in `infer_util`! We should also think about what goes into `util` and `infer_util`. For instance, since we are exposing our version of control flow primitives, it is worth documenting why we have that and how to use it for debugging. Otherwise, it is better to not expose it as part of the API.

Reply:
Another reason for the separation is to just avoid circular imports - `util` shouldn't depend on any other `numpyro` modules, as it currently stands. `infer_util` on the other hand has many other dependencies like `handlers`, `infer`, etc. Right now, it is very easy for internal code to use anything from `util` without worrying about circular imports. On the other hand, utilities in `infer_util` should be mostly for end users (/tutorials/examples), and not as much for internal usage. We have a few that are used internally (e.g. potential_energy) which I think are best moved from `infer_util` to some other module for internal usage. 

Anyways, this PR is the first one. We should think more closely about our organization based on these principles and change it soon, since such changes are backwards incompatible and we will be stuck with them.

Reply:
I just mean to move these files to `infer` folder.

> most users of the library should care about anything in hmc_util, and if they do, it is probably worth exposing it directly in infer_util

agreed!

> For instance, since we are exposing our version of control flow primitives, it is worth documenting why we have that and how to use it for debugging. Otherwise, it is better to not expose it as part of the API.

You are right. It is better to not expose them. Only `control_flow_prims_disabled` is useful for us and for users (to debug).